Improve Tabs indicator animation and related utils

DaniGuardiola opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This PR simplifies the utils used in animating the indicator, which is also the base for the upcoming ToggleGroupControl animation which is done but will be submitted in a follow-up.

Additionally, it switches to a `transform`-based strategy, which should speed up the animation even further.

This work is also related to the `useResizeObserver` refactoring efforts (#64820). Here, the utility is radically simplified (and renamed to `useResizeObserver`) with the goal of unifying with that separate effort in a separate PR.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We had to make some compromises to prevent overflow due to the previous method of measuring positions of elements relative to their offset parent, which resulted in a sub-0.5px error margin (see https://github.com/WordPress/gutenberg/pull/61979). This updated approach has exact precision which eliminates some critical issues in the upcoming ToggleGroupControl animation work.

I made a playground to illustrate this: https://stackblitz.com/edit/solidjs-templates-grt7kw?file=src%2FApp.tsx

The switch to transform provides better animation performance.

The refactor of `useResizeObserver` aims for upcoming unification in relation to #64820.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

See diff, it's all extensively documented.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

Try the tabs indicator animation in the Storybook, plus a few Tabs instances in Gutenberg for good measure (there's a complete list of instances and where to find them in the description of https://github.com/WordPress/gutenberg/pull/64371 if that's useful).

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

Interact with tabs through keyboard.

## Screenshots or screencast <!-- if applicable -->

n/a (just check the storybook, there are no appreciable changes other than the improved subpixel precision)

DaniGuardiola committed:
Refactor utils and switch Tabs indicator animation to `transform`.

ciampo commented on packages/components/src/utils/element-rect.ts:
```diff
@@ -152,28 +150,44 @@ export const NULL_ELEMENT_OFFSET_RECT = {
+	// by the browser.
+	const rect = element.getBoundingClientRect();
+	const offsetParentRect =
+		element.offsetParent?.getBoundingClientRect() ??
```
If the offset parent also was affected by a transform compared to its own offset parent, would that have an impact? In other words, do we need recursion going up the DOM tree?

DaniGuardiola replied:
This is something I thought of, and it just seems like an edge case we just don't have. Even then, it seems like a relative measure between parent and child would be of little use in a context where the parent and the child have different scaling values applied.

DaniGuardiola committed:
docs tweak

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -24,7 +24,7 @@ export const TabListWrapper = styled.div`
 
 	@media not ( prefers-reduced-motion: reduce ) {
```
```suggestion
	@media not ( prefers-reduced-motion ) {
```

DaniGuardiola replied:
good catch!

ciampo reviewed (commented):
Loving the cleanup!

In my tests, the indicator is not showing when switching the writing direction to RTL. We need to RTL-proof some settings, I guess

![Screenshot 2024-08-30 at 18 27 48](https://github.com/user-attachments/assets/dae5ccfe-de2f-4c16-bf43-23876fa067fc)

DaniGuardiola commented:
@ciampo thanks for the review! Do you know if the RTL issue is pre-existing? Because it so, this can be merged and that can be treated as a separate bug. Though I can't really think of a reason RTL would break things, must be missing something 🤔

DaniGuardiola committed:
Update packages/components/src/tabs/styles.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

ciampo commented:
Trunk looks fine to me, so it seems like something introduced by this PR.

![Screenshot 2024-08-31 at 10 51 17](https://github.com/user-attachments/assets/d7f16ba9-6936-487a-ad8c-be2e3cabcace)


I haven't looked deeply into it, but here's where I'd look:

- the first suspect that I see is the `transform-origin` ?
- in general, we need to flip anything that is explicitly or implicitly anchored to the left and scaling to the right
- when the tabs don't "fill" the whole tablist width, that may affect the positioning of the indicator in RTL screen? Maybe we could consider changing from `display: flex` to `display: inline-flex` ?

ciampo commented:
Update: we should look at @t-hamano 's proposed fixes in #64965 as an inspiration for what tweaks to make to support the animation in RTL layouts

t-hamano commented:
Curiously, the visual width of the indicator seems to change depending on the display and browser zoom. Is `transform-origin` or `transform: scaleX()` affecting it? I can't see this issue in trunk.

✅ Display scale: 100%

![image](https://github.com/user-attachments/assets/5f62bae9-6a5b-4be0-9c85-d871f49e6def)

❌ Display scale: 125%

![image](https://github.com/user-attachments/assets/5321768e-eea8-4708-8b98-a3a13446cd00)

❌ Display scale: 100%,  Browser zoom: 150%

![image](https://github.com/user-attachments/assets/9906f4b5-0841-4517-8cae-6a7856a67828)

This issue occurs in Chrome but not in Firefox.

I'm testing on Windows OS, can you reproduce this issue on MacOS?

tyxla commented:
> I'm testing on Windows OS, can you reproduce this issue on MacOS?

Good find @t-hamano 👍 Can reproduce in the latest Chrome on MacOS:

![Screenshot 2024-09-02 at 12 49 01](https://github.com/user-attachments/assets/3b8750b2-55aa-434b-94f8-d4d165bb42ce)

Can't reproduce in `trunk`:

![Screenshot 2024-09-02 at 12 49 54](https://github.com/user-attachments/assets/ac013f94-0e94-4cc5-a622-284830ab73f2)

DaniGuardiola committed:
Add RTL support.

DaniGuardiola commented:
Just added RTL support. Thanks @ciampo and @t-hamano for detecting the issue and @t-hamano for attempting a fix, although now obsolete due to the change to the transform strategy. Your PR will be helpful to me in the follow-up I'm working on for the `ToggleGroupControl`, since that one will need to use the previous positioning-based strategy, so thank you!

Regarding indicator size, I cannot reproduce in any browser, including Chrome in Mac OS. Are y'all in latest? Either way, it seems like a bug to me that's probably just been fixed so I think we'll be fine.

t-hamano commented:
@DaniGuardiola 

> Regarding indicator size, I cannot reproduce in any browser, including Chrome in Mac OS.

Can you reproduce the issue by changing the zoom level of the Chrome browser? My Chrome version is `128.0.6613.114`, but I can reproduce the issue by changing the browser zoom level to something other than 100%.

ciampo commented:
I think it needs a CHANGELOG entry under the "Bug fixes" section since it implicitly fixes the RTL bug flagged by @t-hamano

DaniGuardiola commented:
@t-hamano I'm on the same version and managed to reproduce it under specific zoom levels. Investigating now.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -33,25 +33,38 @@ export const TabListWrapper = styled.div`
+		}
+
+		&:dir( rtl )::after {
+			right: 0;
```
Instead of setting `right` here as a RTL exception, could we use `inset-inline-start` ?

DaniGuardiola replied:
Actually I just realized none are necessary as the pseudo-element naturally falls in the right place by default. So I'll just remove the property 👍

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -33,25 +33,38 @@ export const TabListWrapper = styled.div`
+
+		&:dir( rtl )::after {
+			right: 0;
+			transform-origin: top right;
```
I wonder if there is a logical version to `transform-origin` ? Maybe using numbers/percentages instead of keywords?

DaniGuardiola replied:
That'd be cool but it doesn't seem to be the case. I think it'd need to be a `start` or `end` keyword.

ciampo commented on packages/components/src/utils/element-rect.ts:
```diff
@@ -144,36 +152,56 @@ export type ElementOffsetRect = {
-		// since `offsetWidth` and `offsetHeight` are rounded to the nearest pixel.
-		width: parseFloat( getComputedStyle( element ).width ),
-		height: parseFloat( getComputedStyle( element ).height ),
+		// To obtain the right values for the position:
```
Nit — but just to disambiguate in a context where `right` is used for different meanings

```suggestion
		// To obtain the correct values for the position:
```

DaniGuardiola replied:
Good idea.

DaniGuardiola replied:
Went with "adjusted".

t-hamano commented:
For some reason, the scaling by `transform: scaleX()` doesn't seem to be accurate. Maybe it has to do with subpixel rendering.

I have prepared a playground:

https://codepen.io/Tetsuaki-Hamno/pen/MWMPebv

Both boxes are expected to be 100px wide, but the boxes scaled by `transform: scaleX()` are not visually the correct width due to the browser zoom level.


https://github.com/user-attachments/assets/30ef7aca-d63d-47d7-a807-6ce2d109fd8f

DaniGuardiola committed:
Addressed @ciampo's comments.

t-hamano commented:
One approach to the solution is to replace the base width value with a larger value, but this might not be ideal.

<details><summary>Diff</summary>

```diff
diff --git a/packages/components/src/tabs/styles.ts b/packages/components/src/tabs/styles.ts
index f335a90c02..9c8f70179c 100644
--- a/packages/components/src/tabs/styles.ts
+++ b/packages/components/src/tabs/styles.ts
@@ -43,7 +43,7 @@ export const TabListWrapper = styled.div`
                &::after {
                        bottom: 0;
                        height: 0;
-                       width: 1px;
+                       width: 100px;
index 2977d6a628..a68b662c8e 100644
--- a/packages/components/src/tabs/tablist.tsx
+++ b/packages/components/src/tabs/tablist.tsx
@@ -81,7 +81,7 @@ export const TabList = forwardRef<
                                '--indicator-top': indicatorPosition.top,
                                '--indicator-right': indicatorPosition.right,
                                '--indicator-left': indicatorPosition.left,
-                               '--indicator-width': indicatorPosition.width,
+                               '--indicator-width': indicatorPosition.width / 100,
                                '--indicator-height': indicatorPosition.height,
                                ...otherProps.style,
                        } }
```

</details>

DaniGuardiola commented:
Seems like a bug related to anti-aliasing and rounding: https://stackoverflow.com/a/52159123

I've been trying some ideas like the one in that SO question: https://stackoverflow.com/a/65992479

However, none have been successful so far. Thanks for also looking into it @t-hamano, I think using a larger scale might actually help, since it's an eager rounding issue.

DaniGuardiola committed:
Correct for antialiasing rounding.

DaniGuardiola commented:
Update: found a solution that seems to work correctly in my testing. When changing zoom levels live, sometimes it hangs to the previously computed value which makes it look wrong, however I don't think it's a big deal since as soon as the tabs are interacted with it gets corrected. See last commit.


https://github.com/user-attachments/assets/83bd3956-da45-44aa-9229-d2134fbca3be

DaniGuardiola committed:
Make antialiasing adjustment optional.

DaniGuardiola commented:
Update: this fix causes problems on other browsers that do not exhibit this behavior. A solution might be to find a way that the issue is happening in the first place, and only apply the fix then. This is getting a bit messy so using a larger base width might be the simpler solution after all, and probably good enough for our use case.

DaniGuardiola committed:
Use larger base value and revert antialiasing adjustment code.

DaniGuardiola committed:
DRY RTL

DaniGuardiola commented:
Final (?) update: I ended up going with @t-hamano's suggestion of using a larger base value, and it works great!

Also @ciampo I DRYed up the RTL styling logic, check it out. This is ready for final (?) review!

t-hamano commented on packages/components/src/tabs/stories/index.story.tsx:
```diff
@@ -70,6 +70,15 @@ const Template: StoryFn< typeof Tabs > = ( props ) => {
 
 export const Default = Template.bind( {} );
 
+export const RTL = () => {
+	return (
+		<div dir="rtl" style={ { padding: '1rem' } }>
+			{ /* @ts-expect-error - This is a temporary hack until we migrate to CSF 3 and my sanity is restored. */ }
+			<Template />
+		</div>
+	);
+};
+
```
Since Storybook currently has an option to switch language direction, is this unnecessary?

![a9a889fac500c8c4515d78370e86c5b5](https://github.com/user-attachments/assets/962e6fe7-bc36-428a-8c50-cf3f7a3c31bb)

DaniGuardiola replied:
TIL! Removing the story.

DaniGuardiola replied:
Done

t-hamano commented on packages/components/src/tabs/styles.ts:
```diff
@@ -38,20 +38,41 @@ export const TabListWrapper = styled.div`
+		/* Using a large value to avoid antialiasing rounding issues
+			when scaling in the transform, see: https://stackoverflow.com/a/52159123 */
+		height: 100px;
+		transform: translateY( calc( var( --indicator-top ) * 1px ) )
+			scaleY( calc( var( --indicator-height ) / 100 ) );
```
```suggestion
		transform: translateY( calc( var( --indicator-top ) * 1px ) )
			scaleY( calc( var( --indicator-height ) / 100 ) );
		transform-origin: top;
```

In a vertical layout, I think we need to start from the top. Otherwise, the background color will be misaligned:

![image](https://github.com/user-attachments/assets/3e68058b-72b8-4403-b5ae-f8e7431b0564)

```

DaniGuardiola replied:
We start from the top already. Can't repro this.

t-hamano replied:
Strangely enough, I can reproduce this in my Storybook. The initial value of `transform-origin` is `50% 50% 0`, and `transform-origin` does not seem to be applied 🤔

![tabs](https://github.com/user-attachments/assets/199ede36-b02a-411e-a1da-fc4c37f74cda)

![image](https://github.com/user-attachments/assets/943662eb-fee9-4795-84cb-6a4fdab330b9)

I can also see the problem in the Preferences panel of the block editor:

DaniGuardiola replied:
Are you sure you're in the latest version of the branch? Can you check that you have the `transform-origin: calc( ( 1 - var( --direction-factor ) ) * 50% ) 0%;` like in the styles source? Can you investigate whether it's applied correctly? Depending on the `--direction-factor` variable, the result for the first parameter (X axis) should be 0% (LTR) or 100% (RTL), and the second (Y axis) should always be 0% since it's a static value.

DaniGuardiola replied:
Wait, I managed to repro. Weird, let me take a look.

DaniGuardiola replied:
Alright, it was a dumb mistake I made haha. I placed the origin only in the horizontal tabs styles instead of in the more global scope. Fixed now!

DaniGuardiola committed:
Remove RTL story (redundant since Storybook has a dynamic setting to test RTL).

DaniGuardiola committed:
Fix bug.

t-hamano commented on packages/components/src/tabs/styles.ts:
```diff
@@ -34,24 +34,46 @@ export const TabListWrapper = styled.div`
 
+		--direction-factor: 1;
+		--indicator-start: var( --indicator-left );
+		&:dir( rtl ) {
```
Due to the influence of c347fad, this selector is now applied to the `::after` pseudo-element. As I understand it, we cannot use `:dir()` on the pseudo-element itself. As a result, in RTL languages, `--direction-factor` remains `1` instead of `-1`, and the indicator is not displayed 😅

![direction-factor](https://github.com/user-attachments/assets/21e07f18-e95a-412d-841b-94d6c1959397)

DaniGuardiola replied:
Good catch lol. Fixed!

DaniGuardiola committed:
Fix bug (for real this time).

t-hamano reviewed (approved):
LGTM! In my testing, I found no issues and believe this is the ideal approach at this time. Thank you 👍

- Indicator size, position, and animation
  - ✅ Works correctly in both LTR and RTL languages
  - ✅ Does not shift when changing browser zoom level
  - ✅ Transition effects are disabled when `prefers-reduced-motion` is enabled
- ✅ Displays correctly in Windows high contrast mode
- ✅ Button backgrounds in the vertical tabs are correct
- ✅ The overflow issue fixed in #61979 has not reoccurred.

t-hamano commented:
Finally: As mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/64926#issuecomment-2324367913), I think the changelog needs to be updated.

DaniGuardiola committed:
Add changelog entry.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -22,36 +22,57 @@ export const TabListWrapper = styled.div`
 		content: '';
 		position: absolute;
 		pointer-events: none;
+		transform-origin: calc( ( 1 - var( --direction-factor ) ) * 50% ) 0%;
```
This may be tricky to understand for a developer with fresh eyes — should we add a comment explaining that, based on the writing direction, the x-axis transform origin may be either `0%` or `100%` ?

DaniGuardiola replied:
Good point. Honestly I think I'm just gonna use another variable (something like `direction-x-origin`) and explicitly write the value. Readable > smart.

DaniGuardiola replied:
I did it, let me know what you think.

ciampo reviewed (commented):
Unfortunately, it looks like the indicator is not correct when tabs are rendered inside a popover

https://github.com/user-attachments/assets/f3b4556a-6e8e-4977-b578-49b3fdc842e4

t-hamano commented:
> it looks like the indicator is not correct when tabs are rendered inside a popover

I can't reproduce the issue now 🤔 Is ​​it a browser-specific issue?

DaniGuardiola commented:
can't repro either

(╯°□°）╯︵ ┻━┻

DaniGuardiola committed:
De-cleverfy code.

ciampo commented:
Let me try to nuke my local env, rebuild everything, install fresh WP, and test again.

ciampo commented:
Tested again on a fresh install — I can continue to reproduce on Chrome, Firefox and Safari MacOs 🤷 

@WordPress/gutenberg-components am I the only one?

t-hamano commented:
The latest commit hash for this PR is 4db2b351ce, is the issue occurring as of this commit?

ciampo commented:
Yup, that's the commit. 

Are you both testing changes on Windows, or is @DaniGuardiola testing on MacOS? Just trying to find what could be the difference.

https://github.com/user-attachments/assets/38fd0b28-6718-4563-a476-dc8b1e823bec

DaniGuardiola commented:
@ciampo tested in Mac.

t-hamano commented:
Can you reproduce this issue in Playground? If you enter the PR number, we all should be able to test it in the same environment.

https://playground.wordpress.net/gutenberg.html


https://github.com/user-attachments/assets/da6a02f1-1ac8-4e10-a52b-6d42b77cf57d

DaniGuardiola commented:
Tested on Windows now using the playground, still can't repro.

ciampo commented:
I can 🤷 


https://github.com/user-attachments/assets/3925c419-324f-4e08-9f82-acfd0250cda6

ciampo commented:
I also tested on latest chrome on MacOS Sequoia via browserstack (free trial), and still managed to repro 🤷 

![Screenshot 2024-09-02 at 21 32 02](https://github.com/user-attachments/assets/7216c942-6720-49e7-be36-a6bbfda9308a)

DaniGuardiola commented:
HOW

tyxla commented:
Very easy to repro with the latest Chrome in macOS:

https://github.com/user-attachments/assets/596782cd-5cda-4a65-b3c7-28a36db29baf

DaniGuardiola commented:
AH! Needed to set a gradient, I was missing that. I'm onto something now...

![image](https://github.com/user-attachments/assets/e8a38f2e-4537-48a3-bb15-93b17aab145d)

t-hamano commented:
I looked into the cause of the problem and found that when the `Tabs` component is rendered in the `Popover` component, `rect.width` and `rect.height` becomes zero [here](https://github.com/WordPress/gutenberg/pull/64926/files#diff-5a5a740f296a0fb34ea44283d6e5ab1d4a96f8af6ac06db3dd003681c8d99765R179).

This may lead to some CSS variables becoming `NaN` as a result.

Perhaps the problem is related to the animation. If I enable `prefers-reduced-motion` to disable the popover animation, the issue does not occur. Is there any way to solve this issue..?

https://github.com/user-attachments/assets/6abbef54-20f4-4d9a-b3fc-fb8cad888243

DaniGuardiola committed:
Sync useResizeObserver with #64943 and make useTrackElementOffsetRect resilient.

DaniGuardiola commented:
I've updated the PR with two details:

- Synced the `useResizeObserver` hook with the latest version from #64943, as the ultimate goal is to remove it from here and use it from `@wordpress/compose`.
- Added a system to `useTrackElementOffsetRect` for resilience to an edge case (like the one we're running into with the background color popover) by attempting to measure again after a frame, and if that fails, polling until it succeeds. Thanks @ciampo for pairing with me to look into this, and @t-hamano for helping out as well.

If you look at the popover now, you'll see that it's properly sized. In this case, the first fallback (requestAnimationFrame) succeeds. Polling is really just a last resort fallback, in case there's a similar case where there's also an animation delay. It should be really rare, if it ever happens at all, but at least we can rest assured it will never break the indicator. It also should not be too heavy on the CPU, 10 measurements seems like a good way to balance performance with speed so that there's no significant flashes or delays for the user.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/tabs-indicator-animation-and-utils

DaniGuardiola committed:
Deduplicate utility and clean up.

DaniGuardiola commented:
Update: I have now removed the duplicated utility since https://github.com/WordPress/gutenberg/pull/64943 has been merged.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -22,36 +22,59 @@ export const TabListWrapper = styled.div`
+							1px
+					)
+				)
+				scaleX( calc( var( --indicator-width ) / 100 ) );
```
Nit: if the `width: 100px` and the division by `100` are related because of the anti-aliasing fix, we could extract `100` to a CSS variable to better show what's going on.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -22,36 +22,59 @@ export const TabListWrapper = styled.div`
+			when scaling in the transform, see: https://stackoverflow.com/a/52159123 */
+		height: 100px;
+		transform: translateY( calc( var( --indicator-top ) * 1px ) )
+			scaleY( calc( var( --indicator-height ) / 100 ) );
```
Same here re. potentially extracting the `100` anti-aliasing factor

ciampo reviewed (approved):
LGTM 🚀 

Changes tested well on my machine, both in Storybook and in the editor, LTR and RTL, horizontal and vertical, automatic or manual tab activation, in sidebars and in popovers.

Left a couple of minor comments, but they're not blocking.

Also, this needs an updated CHANGELOG entry (likely under "internal", since `Tabs` is still private)

DaniGuardiola committed:
DRY antialiasing factor.

DaniGuardiola committed:
Changelogs.
